tests/conftest.py

```diff
@@ -55,9 +56,141 @@ def pytest_addoption(parser):
     )
 
 
+def _remove_ports(ports, file, output_file):
+    command = ["sed"] + [port for port in ports for port in ["-e", f"/{port}/d"]]
+    with open(output_file, "w") as f:
+        subprocess.run(command + ["-e", "/ports:/d", file], stdout=f)
+
+
+@pytest.fixture(scope="session", autouse=True)
+def modify_compose_for_testing(request):
+    demo = "../docker-compose.demo.yml"
+    storage_proxy_demo = "../storage-proxy/docker-compose.storage-proxy.demo.yml"
+    # Output files after modifying for testing
+    testing = "../docker-compose.testing.yml"
+    storage_proxy_testing = "../storage-proxy/docker-compose.storage-proxy.testing.yml"
+
+    # Published ports in docker-compose-demo and docker-compose.storage-proxy.demo
+    demo_ports = ["9000:9000", "8080:8080", "443:443", "80:80"]
+    demo_storage_proxy_ports = ["9000:9000"]
+
+    # Remove published ports from docker-compose-demo
+    _remove_ports(demo_ports, demo, testing)
+    # Remove published ports from docker-compose.storage-proxy.demo
+    _remove_ports(demo_storage_proxy_ports, storage_proxy_demo, storage_proxy_testing)
+
+    # Disable download speed limits
+    subprocess.run(["sed", "-e", "s/DOWNLOAD_SPEED/#DOWNLOAD_SPEED/", "-i", testing])
+
+    # Whitelist all IPs/DNS names in the gateway
+    subprocess.run(
+        ["sed", "-e", "s/ALLOWED_HOSTS: .*/ALLOWED_HOSTS: ~./", "-i", testing]
+    )
+
+
+def _extract_fs_from_image(request, client_compose_file):
+
+    d = "/output"
+
+    def cleanup():
+        shutil.rmtree(d, ignore_errors=True)
+
+    request.addfinalizer(cleanup)
+
+    image_type = "mender-client"
+    if "gateway" in client_compose_file:
+        image_type = "mender-gateway"
+
+    with open(client_compose_file) as f:
+        data = yaml.safe_load(f)
+        try:
+            image = data["services"][image_type]["image"]
+        except Exception as e:
+            raise ValueError(f"Key not found in {client_compose_file}: {e}")
+
+    subprocess.run(
+        [
+            "docker",
+            "run",
+            "--rm",
+            "--privileged",
+            "--entrypoint",
+            "/extract_fs",
+            "--volume",
+            d + ":/output",
+            image,
+        ]
+    )
+
+    return d
+
+
+def _image(request, compose_file, filename):
+    dirname = _extract_fs_from_image(
+        request, os.path.join(THIS_DIR, "..", compose_file)
```

Comment:
Q: AFAIK all the compose files are in the same directory, and we are only extracting the path. Is the compose_file parameter needed then?

Reply:
I would say it's needed, since _image passes it to the extract fs function – which needs it to extract the filesystem.

Reply:
Ohh, now I see it. I had to stare at this one for a while actually :laughing: There's a function call there!